ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,89 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+class ImpendingTimeout(Exception):
+    pass
+
+
+def crash_flush(_, __):
+    """Raises an `ImpendingTimeout` error which causes the tracer to shutdown."""
+    # TODO: Update to tag error in local root parent instead of raising the error.
+
+    raise ImpendingTimeout("Datadog detected an Impending Timeout")
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE", 0))
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        remaining_time_in_millis = apm_flush_deadline
+
+    remaining_time_in_millis = (remaining_time_in_millis - 100) / 1000
+    signal.setitimer(signal.ITIMER_REAL, remaining_time_in_millis, 0)
+
+
+def datadog(func, args, kwargs):
```

Comment:
Maybe we could make the function name more descriptive?

Reply:
Let me know if `datadog_instrumentation` is descriptive enough